### ISSUE TITLE
Error finding timestamps for builds

### ISSUE THREAD

**Issue by forever84721** [jenkinsci/build-time-blame-plugin#27]
**Describe the bug**
I followed the README steps, When clicking "Reprocess Report"
![image](https://user-images.githubusercontent.com/22338037/91789547-6df4a380-ec41-11ea-8b93-124498b517ca.png)

**Screenshots**
![image](https://user-images.githubusercontent.com/22338037/91789727-cd52b380-ec41-11ea-8f9a-9c20ac4b2597.png)
![image](https://user-images.githubusercontent.com/22338037/91789748-d80d4880-ec41-11ea-9ec9-59c73110e9c8.png)
![image](https://user-images.githubusercontent.com/22338037/91789823-fffcac00-ec41-11ea-8a7c-8ede4600f44c.png)

**Additional context**
Jenkins 2.235.5
Build Time Blame Plugin 2.0.0
Timestamper 1.11.5

**Comment by LTegtmeier** [comment]
Thanks for reporting this! I'll try to recreate this on a test job. 

Do you get the error on all jobs on this Jenkins Master or just this one?

**Comment by forever84721** [comment]
It didn't display error in another job, but the report didn't work too.

**Comment by cmchoi2000** [comment]
I have same error. Error finding timestamps for builds: 
My log is 28MB and it has lines in excess of 8128 characters.

Jenkins 2.249.1
Build Time Blame Plugin 2.0.0
Timestamper 1.11.5

**Comment by LTegtmeier** [comment]
The fix from @karkusviktor was released in [v2.0.1](https://github.com/jenkinsci/build-time-blame-plugin/releases/tag/build-time-blame-2.0.1).

**Comment by forever84721** [comment]
I updated to v2.0.1, but It just didn't display error, the report didn't work too.
Do I need to do anything after the update?

**Comment by LTegtmeier** [comment]
~~If you do not get an error message, it is likely a different problem.~~

~~Can you file a separate issue with details like your Jenkins version, timestamper version, and your console output?~~

I thought you weren't the original poster. You get a blank report on these same builds that failed before?

Keep in mind that the plugin only considers successful builds where timestamps were enabled. I see you only configured it to look at two builds. Are your most recent builds successes?

**Comment by forever84721** [comment]
In version 2.0.0
There is a project that will display "Error finding timestamps for builds" regardless of whether the build succeeds or fails
The other project is blank regardless of whether the build succeeds or fails, and no errors are displayed.
After updating 2.0.1
Now regardless of the success or failure of the build, it is blank and no errors are displayed.

Can it work when you use it? Maybe it’s a problem with my environment?
Thanks.

**Additional context**
Jenkins 2.249.1
Build Time Blame Plugin 2.0.1
Timestamper 1.11.5

My recent builds are successes
This is my output
00:00:00.001 Started by timer
00:00:00.001 Running as SYSTEM
00:00:00.001 Building in workspace C:\Users\Administrator\AppData\Local\Jenkins.jenkins\workspace\CentralFileStorage Angular
00:00:00.002 The recommended git tool is: NONE
00:00:00.002 using credential 6ec9fb75-edb7-4edb-b689-0f90d1f7a148
00:00:00.467  > C:\Program Files\Git\bin\git.exe rev-parse --is-inside-work-tree # timeout=10
00:00:00.487 Fetching changes from the remote Git repository
00:00:00.487  > C:\Program Files\Git\bin\git.exe config remote.origin.url http://192.168.20.11:3000/Jay/CentralFileStorage.git # timeout=10
00:00:00.508 Fetching upstream changes from http://192.168.20.11:3000/Jay/CentralFileStorage.git
00:00:00.508  > C:\Program Files\Git\bin\git.exe --version # timeout=10
00:00:00.531  > git --version # 'git version 2.28.0.windows.1'
00:00:00.532 using GIT_ASKPASS to set credentials 
00:00:00.534  > C:\Program Files\Git\bin\git.exe fetch --tags --force --progress -- http://192.168.20.11:3000/Jay/CentralFileStorage.git +refs/heads/*:refs/remotes/origin/* # timeout=10
00:00:01.260  > C:\Program Files\Git\bin\git.exe rev-parse "refs/remotes/origin/master^{commit}" # timeout=10
00:00:01.285  > C:\Program Files\Git\bin\git.exe rev-parse "refs/remotes/origin/origin/master^{commit}" # timeout=10
00:00:01.315 Checking out Revision abe42f21cec23a9cdb0c880d58b25e013ebc682b (refs/remotes/origin/master)
00:00:01.315  > C:\Program Files\Git\bin\git.exe config core.sparsecheckout # timeout=10
00:00:01.337  > C:\Program Files\Git\bin\git.exe checkout -f abe42f21cec23a9cdb0c880d58b25e013ebc682b # timeout=10
00:00:01.549 Commit message: "Update delete file url"
00:00:01.550  > C:\Program Files\Git\bin\git.exe rev-list --no-walk abe42f21cec23a9cdb0c880d58b25e013ebc682b # timeout=10
00:00:01.595 [CentralFileStorage Angular] $ cmd /c call C:\Users\ADMINI1\AppData\Local\Temp\jenkins2869904059452739686.bat
00:00:01.607 
00:00:01.607 C:\Users\Administrator\AppData\Local\Jenkins.jenkins\workspace\CentralFileStorage Angular>npm i 
00:00:07.671 npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@2.1.3 (node_modules\fsevents):
00:00:07.671 npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})
00:00:07.676 npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.13 (node_modules\webpack-dev-server\node_modules\fsevents):
00:00:07.677 npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.13: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})
00:00:07.681 npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.13 (node_modules\watchpack-chokidar2\node_modules\fsevents):
00:00:07.681 npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.13: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})
00:00:07.681 
00:00:07.776 removed 1 package and audited 1466 packages in 5.506s
00:00:08.218 
00:00:08.218 61 packages are looking for funding
00:00:08.218   run `npm fund` for details
00:00:08.218 
00:00:08.219 found 0 vulnerabilities
00:00:08.219 
00:00:08.298 [CentralFileStorage Angular] $ cmd /c call C:\Users\ADMINI1\AppData\Local\Temp\jenkins3739628989601773319.bat
00:00:08.310 
00:00:08.310 C:\Users\Administrator\AppData\Local\Jenkins.jenkins\workspace\CentralFileStorage Angular>npm run build -- --prod --output-path "C:\IIS Application\CentralFileStorage" --base-href /CentralFileStorage/ 
00:00:08.787 
00:00:08.787 > central-file-storage@0.0.0 build C:\Users\Administrator\AppData\Local\Jenkins.jenkins\workspace\CentralFileStorage Angular
00:00:08.787 > ng build --prod --base-href /CentralFileStorage/ "--prod" "--output-path" "C:\IIS Application\CentralFileStorage" "--base-href" "/CentralFileStorage/"
00:00:08.787 
00:00:36.308 Generating ES5 bundles for differential loading...
00:00:36.655 ES5 bundle generation complete.
00:00:36.656 
00:00:36.656 chunk {4} polyfills-es5.41a50de0451a8d01b801.js (polyfills-es5) 129 kB [initial] [rendered]
00:00:36.656 chunk {3} polyfills-es2015.578ae4c771d50a6ed3ee.js (polyfills) 36.1 kB [initial] [rendered]
00:00:36.656 chunk {7} 7-es2015.b8b0abf253e8077875dc.js () 2.56 kB  [rendered]
00:00:36.656 chunk {7} 7-es5.b8b0abf253e8077875dc.js () 3.07 kB  [rendered]
00:00:36.656 chunk {6} 6-es2015.b99206d62a57e4123953.js () 1.07 kB  [rendered]
00:00:36.656 chunk {6} 6-es5.b99206d62a57e4123953.js () 1.46 kB  [rendered]
00:00:36.656 chunk {8} 8-es2015.09b96a7975a77996aa84.js () 992 bytes  [rendered]
00:00:36.656 chunk {8} 8-es5.09b96a7975a77996aa84.js () 1.33 kB  [rendered]
00:00:36.656 chunk {1} 1-es2015.0549069db7419acb39c7.js () 151 kB  [rendered]
00:00:36.656 chunk {1} 1-es5.0549069db7419acb39c7.js () 169 kB  [rendered]
00:00:36.656 chunk {2} main-es2015.bdc28fadd2d28099635e.js (main) 497 kB [initial] [rendered]
00:00:36.656 chunk {2} main-es5.bdc28fadd2d28099635e.js (main) 575 kB [initial] [rendered]
00:00:36.656 chunk {0} runtime-es2015.a0da28269de949042704.js (runtime) 2.33 kB [entry] [rendered]
00:00:36.656 chunk {0} runtime-es5.a0da28269de949042704.js (runtime) 2.33 kB [entry] [rendered]
00:00:36.656 chunk {9} 9-es2015.71f61416f24d986aebbb.js () 998 bytes  [rendered]
00:00:36.656 chunk {9} 9-es5.71f61416f24d986aebbb.js () 1.34 kB  [rendered]
00:00:36.656 chunk {5} styles.20901389a334c199bb71.css (styles) 66.4 kB [initial] [rendered]
00:00:36.656 Date: 2020-09-27T16:00:37.914Z - Hash: 52df7ebcc60f2bac57fd - Time: 25049ms
00:00:36.657 
00:00:36.657 
00:00:36.810 [CentralFileStorage Angular] $ cmd /c call C:\Users\ADMINI1\AppData\Local\Temp\jenkins1121281324468181513.bat
00:00:36.823 
00:00:36.823 C:\Users\Administrator\AppData\Local\Jenkins.jenkins\workspace\CentralFileStorage Angular>move /y "C:\IIS Application\CentralFileStorage\assets\web.config" "C:\IIS Application\CentralFileStorage" 
00:00:36.824         1 file(s) moved.
00:00:36.824 
00:00:36.824 C:\Users\Administrator\AppData\Local\Jenkins.jenkins\workspace\CentralFileStorage Angular>exit 0 
00:00:37.451 [Office365connector] Matched status 'SUCCESS' for webhook with name 'Bot'.
00:00:37.452 Finished: SUCCESS

**Comment by LTegtmeier** [comment]
### Background

I'm not sure how you could get an empty report without errors . The report first checks for builds that are completed successfully. Then, it stops if it has processed too many builds. Then, it tries to build the report and adds to the failure header if anything goes wrong. 
https://github.com/jenkinsci/build-time-blame-plugin/blob/72721a24f3611375f1100b57cbabbc2d366061bb/src/main/groovy/org/jenkins/ci/plugins/buildtimeblame/action/BlameAction.groovy#L114-L127

### Recreating

I cannot recreate that behavior. I don't regularly use Jenkins anymore and I haven't gotten my test setup working properly yet. I can see a populated report for builds with timestamps. My fresh freeStyleJob builds get timestamps but, new pipeline builds don't have timestamps. The key there is that for all those builds, I get an error header because the timestamps couldn't be found. 
![image](https://user-images.githubusercontent.com/6852365/94508006-a3f05d80-01d6-11eb-8590-b4ab3873357d.png)

I get the same output using the Jenkins and timestamper versions this project builds with or with the versions you posted. 

#### Debugging

* Is the whole report blank or just the graphic? Can you see the other summary tables?
![image](https://user-images.githubusercontent.com/6852365/94508710-2cbbc900-01d8-11eb-8eed-d98d6ae0e633.png)
* Can you try downgrading the timestamper plugin to 1.8.7?
* Does it do anything different with Freestyle jobs vs Jenkins 2.0 Pipeline jobs?

**Comment by karkusviktor** [comment]
If I can add something:

- The direct output which is used by the plugin is more useful for debugging than the console log so like this: "http://localhost:8080/jenkins/job/test-pipeline/1/timestamps/?appendLog&elapsed=S" It should look similar to the console output, but instead of "00:00:00.001 Started by timer" it will look like "1  Started by timer", but it's not dependent on how the user set the plugin (it might display wall clock instead of elapsed time in the Console log) and it's sure that the right plugin is installed so not a really old one which doesn't support such query which is used in this plugin.
- But the best, at least how I did the debugging for myself, was just zip a build log and add it as a log for a dummy job in the dev environment. So zip e.g. /path/to/jenkins/jobs/test-pipeline/builds/1 and then replace the log of a run in the dev env so you can run the debugger against it.

@forever84721 can you attach such a log? I mean the zipped one bc. the first one can be created from a console log. Or maybe that is useful as well if you can't attach the zip, because it is different from the console log in whitespaces.

Update: I just discovered that an extra whitespace is added to the beginning of each line. So if you use '^' all of your regexes then it won't match anything (except ofc. if you prepare the regex for this extra whitespace bug). The reason of this extra whitespace is that between the timestamp and log line in the output of timestamper there are two spaces, so the split arrays second piece (so split[1]) will be always and empty string which will add a whitespace when the code tries to rebuild the original line without the timestamp: https://github.com/jenkinsci/build-time-blame-plugin/blob/72721a24f3611375f1100b57cbabbc2d366061bb/src/main/groovy/org/jenkins/ci/plugins/buildtimeblame/analysis/TimedLog.groovy#L23

Try adding this to LogParserTest, it will pass, **note the extra spaces at the assertion** and in the "Finished" regexp:
```
def 'testFullLog'() {
given:
def build = Mock(Run)
def log = "8  Started by user unknown or anonymous\n" +
        "9  Running as SYSTEM\n" +
        "21  Building in workspace /home/viktor/IdeaProjects/build-time-blame-plugin/work/workspace/test\n" +
        "117  Finished: SUCCESS"
setupMockTimestamperLog(build, Arrays.stream(log.split('\n')).map({ line -> TimedLog.fromText(line) }).collect(Collectors.toList()))
def timestamps = List.of(8,117)

def logParser = new LogParser([
        new RelevantStep(~".*Started by.*", 'Job Started on Executor', true),
        new RelevantStep(~"^ Finished: (SUCCESS|UNSTABLE|FAILURE|NOT_BUILT|ABORTED)\$.*", 'Finished', true),
])

when:
BuildResult buildResult = logParser.getBuildResult(build)

then:
buildResult.build == build
buildResult.consoleLogMatches == [
        buildLogResult('Job Started on Executor', ' Started by user unknown or anonymous', timestamps, 0, 1),
        buildLogResult('Finished', ' Finished: SUCCESS', timestamps, 1, -1),
]
}
```

**Comment by LTegtmeier** [comment]
> The direct output which is used by the plugin is more useful for debugging than the console log so like this: "http://localhost:8080/jenkins/job/test-pipeline/1/timestamps/?appendLog&elapsed=S"

@forever84721, a zip of the full build directory would help the most. But, if you can't get that, the output of that URL is the next best thing. I'm not sure why I forgot about that...

> I just discovered that an extra whitespace is added to the beginning of each line.

@karkusviktor, nice find!

I think that behavior is different with different versions or configurations of the timestamper plugin since it is still working with `^` in my setup.

I created https://github.com/jenkinsci/build-time-blame-plugin/pull/29 since I think that would address that issue. I'll TDD that fix, look for a cleaner way, and work on a release tonight (CST).

**Comment by forever84721** [comment]
[build-time-blame-plugin-issues27.txt](https://github.com/jenkinsci/build-time-blame-plugin/files/5299172/build-time-blame-plugin-issues27.txt)
This?
Thanks for your help.

**Comment by LTegtmeier** [comment]
> This?

Yes! Given that includes the double space character like @karkusviktor pointed out, I think that's your issue. 

As a workaround, can you update your report configuration to not include `^` on the Finished match?

**Comment by forever84721** [comment]
I tried many kinds of regular expressions, but they didn't work.
Can you check it directly for me if possible?
http://61.31.169.97:9999/
Test/test

**Comment by forever84721** [comment]
I see it works!
Thank you very much.
What has changed?

**Comment by karkusviktor** [comment]
@forever84721 Look at it now.
The regex has to match **the whole line** so I reverted back the "Started" regex to the original one and now you have nice graph.
Reverting the finished regex is not necessary as these mark the **beggining** of a step and so the "fininshed" step will take 0 time anyways as it is the last line so it won't be displayed.

**Comment by forever84721** [comment]
I used original regex(.*Started by.*) in version 2.0.0, but it showed "Error finding timestamps for builds", so I changed it to "Running as SYSTEM" for test.
After upgrading to 2.0.1, I did not change back to original regex to test, So is the problem fixed by version 2.0.1?
Thank you!

**Comment by LTegtmeier** [comment]
Leaving this open since the problem with `^` regexes and therefore the defaults is still there.

**Comment by LTegtmeier** [comment]
The fix for `^` regexes was deployed as v2.0.2. Thanks again for finding the issue @karkusviktor!
